examples/hardware_decoding/main.go

```diff
@@ -0,0 +1,190 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/asticode/go-astiav"
+)
+
+var (
+	input       = flag.String("i", "", "the input path")
+	device_type = flag.String("d", "", "the hardware device type like: cuda")
+)
+
+type stream struct {
+	decCodec        *astiav.Codec
+	decCodecContext *astiav.CodecContext
+	hwDeviceContext *astiav.HardwareDeviceContext
+	inputStream     *astiav.Stream
+}
+
+func hw_decoder_init(ctx *astiav.CodecContext, t astiav.HardwareDeviceType) (*astiav.HardwareDeviceContext, error) {
+	hdc, err := astiav.CreateHardwareDeviceContext(t, "", &astiav.Dictionary{})
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create hardware device context: %w", err)
+	}
+	ctx.SetHardwareDeviceContext(hdc)
+	return hdc, nil
+}
+
+func main() {
+	// Handle ffmpeg logs
+	astiav.SetLogLevel(astiav.LogLevelDebug)
+	astiav.SetLogCallback(func(l astiav.LogLevel, fmt, msg, parent string) {
+		log.Printf("ffmpeg log: %s (level: %d)\n", strings.TrimSpace(msg), l)
+	})
+
+	// Parse flags
+	flag.Parse()
+
+	// Usage
+	if *input == "" || *device_type == "" {
+		log.Println("Usage: <binary path> -d <device type> -i <input path>")
+		return
+	}
+
+	hw_device := astiav.FindHardwareDeviceTypeByName(*device_type)
+	if hw_device == astiav.HardwareDeviceTypeNone {
+		log.Fatal(errors.New("main: hardware device not found"))
+	}
+
+	// Alloc packet
+	pkt := astiav.AllocPacket()
+	defer pkt.Free()
+
+	gpu_f := astiav.AllocFrame()
+	defer gpu_f.Free()
+
+	sw_f := astiav.AllocFrame()
+	defer sw_f.Free()
+
+	// Alloc input format context
+	inputFormatContext := astiav.AllocFormatContext()
+	if inputFormatContext == nil {
+		log.Fatal(errors.New("main: input format context is nil"))
+	}
+	defer inputFormatContext.Free()
+
+	// Open input
+	if err := inputFormatContext.OpenInput(*input, nil, nil); err != nil {
+		log.Fatal(fmt.Errorf("main: opening input failed: %w", err))
+	}
+	defer inputFormatContext.CloseInput()
+
+	// Find stream info
+	if err := inputFormatContext.FindStreamInfo(nil); err != nil {
+		log.Fatal(fmt.Errorf("main: finding stream info failed: %w", err))
+	}
+
+	var hw_pix_fmt astiav.PixelFormat
+
+	// Loop through streams
+	streams := make(map[int]*stream) // Indexed by input stream index
+	for _, is := range inputFormatContext.Streams() {
+		var err error
+
+		// Only process video
+		if is.CodecParameters().MediaType() != astiav.MediaTypeVideo {
+			continue
+		}
+
+		// Create stream
+		s := &stream{inputStream: is}
+
+		// Find decoder
+		if s.decCodec = astiav.FindDecoder(is.CodecParameters().CodecID()); s.decCodec == nil {
+			log.Fatal(errors.New("main: codec is nil"))
+		}
+
+		// Alloc codec context
+		if s.decCodecContext = astiav.AllocCodecContext(s.decCodec); s.decCodecContext == nil {
+			log.Fatal(errors.New("main: codec context is nil"))
+		}
+		defer s.decCodecContext.Free()
+
+		hw_pix_fmt, err = astiav.FindSuitableHardwareFormat(s.decCodec, hw_device)
```

Comment:
like this ? 

```go
codec_hardware_configs, err := s.decCodec.HardwareConfigs(hw_device)
if err != nil {
	log.Fatal(fmt.Errorf("main: no codec hardware config found: %w", err))
}

found := false
for _, p := range codec_hardware_configs {
	if p.MethodFlags().Has(astiav.CodecHardwareConfigMethodHwDeviceCtx) && p.HardwareDeviceType() == hw_device {
		hardwarePixelFormat = p.PixelFormat()
		found = true
		break
	}
}

if !found {
	log.Fatal(fmt.Errorf("main: Decoder %s does not support device type", hw_device.String()))
}
```

Reply:
Exactly 👍 

You can even simplify by setting `hardwarePixelFormat` to `astiav.PixelFormatNone` in the first place. That way you don't need the `found` variable and instead of testing `!found` you can test `hardwarePixelFormat == astiav.PixelFormatNone`

Reply:
ok cool